### DISCUSSION TITLE
Minimum Node.js version

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#17797] [Announcements]
History: #800 > #1197 (adding engines field)

Starting with nuxt 3 and bridge, we started warning users to use either the Latest LTS or non-LTS version of Node.js listed in https://nodejs.org/en/. This way we can ensure that we are not locked to old dependencies because of Node.js version support and also not resolving issues fixed in Node.js upstream with workarounds.

The current constraint is: `^14.16.0 || ^16.11.0 || ^17.0.0` 

Node releases: https://nodejs.org/en/about/releases/

Currently, we use two methods for version checks
- Using [engines.node](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#engines) field 
    - Package managers (`npm` and `yarn`) hard fail if using an incompatible Node.js version
    - For `yarn`, can be bypassed by `yarn --ignore-engines`
    - For `npm`, can be bypassed by using `npm install --force` or setting `engine-strict: false` in `.npmrc`
- Using a soft warning in `nuxi` CLI

Open questions:

- Is enforcing using `engines.node` in nuxt packages the best idea? 
   - Does ignore methods solve issue on all platforms (including CI/CD and Sandboxes)
- Should we also add `engines.node` to starter's `package.json`?
- How often we should update semver-**minor** in constraint?
    - We need to check with platforms (table below) to see how often they do this
- How often we should update semver-**major** in constraint? 
    - Always enforcing **latest LTS** as a minimum or all LTS (14+)?  
    - Should it also be alongside with a major version update for nuxt framework packages or a minor version?
- Should we have a relaxed check with engines and more strict warning-only check with `nuxi`?

Reference. Bellow listing platform docs for Node.js version:

Platform | Default  | Supported versions |  Docs
-|-|-|-
Stackblitz| 14.16.x  | ? | ?
Codesandbox | 12.x | 10.x, 12.x, 14.17.x | https://codesandbox.io/docs/configuration
Vercel | 14.x | 14.x | https://vercel.com/docs/runtimes#official-runtimes/node-js
Netlify | ? | [all using nvm] | https://docs.netlify.com/configure-builds/manage-dependencies/#node-js-and-javascript
Github Actions | 14 (latest LTS pulled) | [all using nvm] | https://github.com/actions/setup-node
Google App Engine | ? | 14.x and 16.x | https://cloud.google.com/appengine/docs/standard/nodejs/runtime
Azure web app | ? | ? | https://docs.microsoft.com/en-us/azure/app-service/configure-language-nodejs
Cloudflare pages | 12.18.0 | [all using nvm] | https://developers.cloudflare.com/pages/platform/build-configuration#language-support-and-tools

**Comment by tobiasdiez** [comment]
Azure functions support only Node 14  (see https://docs.microsoft.com/en-us/azure/azure-functions/functions-versions?tabs=in-process%2Cv4&pivots=programming-language-csharp#languages) and Azure static web apps only Node 12 with v 14 and 16 pending (https://github.com/Azure/static-web-apps/issues/623).

**Reply by nogic1008** [reply]
Azure Static Web Apps supports Node 14 and 16 (preview) now
https://docs.microsoft.com/azure/static-web-apps/configuration#selecting-the-api-language-runtime-version

**Comment by ennioVisco** [comment]
@pi0 are you sure node 14 is still supported? I'm getting weird parse errors on components' HTML templates.

**Reply by danielroe** [reply]
I think that is an npm issue. Try using npm version 8.

**Reply by ennioVisco** [reply]
I apologize, it seems it was not an issue of the specific Node/npm version. I created a new issue that tries to explain the problem here: https://github.com/nuxt/framework/issues/3924.
